### PR TITLE
Resolve issues #2, #12, #13, #14: vault_id required, parent retrieval, ingestion integrity, embedding migration

### LINKED ISSUE #2: Track: Harrier Migration - Tech Debt & Remaining Issues
## Overview
This issue tracks remaining tech debt and improvements identified during the comprehensive swarm-mode PR review of the Harrier embedding migration (PR #1).

**Migration Status**: ✅ Core implementation approved for merge. Reranking fully operational and wired.

---

### 1. **Embedding Dimension Migration Path** (MEDIUM Priority)
**Problem**: Existing LanceDB vectors from BGE-M3 migration are incompatible with new Harrier 1024-dim embeddings. No migration script provided.

**Impact**: Existing documents won't be searchable until re-embedded. Users must re-upload documents or run a re-embedding job.

**Action Items**:
- [ ] Document embedding dimension migration requirement in README upgrade section
- [ ] Add note to release.md about re-embedding requirement for existing deployments
- [ ] (Optional) Create migration script to re-embed existing documents: `scripts/migrate_embeddings.py`
- [ ] Add health check warning if documents exist but may be stale embeddings

**Acceptance Criteria**:
- Users understand they must re-upload/re-embed documents when upgrading
- Clear guidance in upgrade documentation
- No silent failures (documents return empty results)

---

### 2. **Sparse Embedding Schema Column - Unused Storage** (LOW Priority - Technical Debt)
**Problem**: Schema retains `sparse_embedding` column for backward compatibility (Option A), but it's never populated post-migration.

**Impact**: Storage bloat (~8-16 bytes per chunk, negligible but clutters schema). No functional impact.

**Action Items**:
- [ ] Document this as planned cleanup in v2.1 or later
- [ ] Consider Option B later: drop column entirely with migration (breaking change for rollback)
- [ ] Add schema validation test to flag unused columns

---

### 3. **Reranker Service Documentation** (LOW Priority)
**Problem**: Reranker service (BAAI/bge-reranker-v2-m3 via TEI) is fully operational, but relationship to embedding/retrieval pipeline could be clearer.

**Impact**: None - reranking is working. Documentation improvement only.

**Action Items**:
- [ ] Add diagram to README showing retrieval → reranking pipeline
- [ ] Document reranking tuning parameters (RERANKER_TOP_N, INITIAL_RETRIEVAL_TOP_K)
- [ ] Add reranking troubleshooting section to admin-guide.md

**Acceptance Criteria**:
- Users understand reranking is enabled by default
- Performance tuning parameters are documented
- Troubleshooting guide covers reranking failures (TEI endpoint down, model loading, etc.)

---

### A. **Hybrid Alpha Simplification**
Pre-migration code had dynamic `effective_alpha` calculation (adjusted based on available search types). Now fixed at `hybrid_alpha` config value.

**Status**: Acceptable tradeoff identified. No action required unless search quality regresses.

**Monitoring**: Track search quality metrics post-deployment.

---

### B. **TEI Service Startup Ordering**
Docker Compose dependencies ensure reranker/harrier-embed healthchecks pass before knowledgevault starts. Order is correct, but monitor for timing issues in slow environments.

**Status**: Verified. No action required unless deployment fails in edge cases.

---

### 1. **Harrier Embedding Dimension Validation**
- [ ] Add test: `test_harrier_embedding_dimension()` - verify 1024-dim output from TEI
- [ ] Add test: `test_embedding_batch_dimension_consistency()` - all embeddings are 1024-dim

### 2. **TEI Service Communication**
- [ ] Add test: `test_harrier_embed_availability()` - mock TEI health check
- [ ] Add test: `test_harrier_embed_openai_format()` - verify OpenAI-compatible format

### 3. **Docker Compose Orchestration**
- [ ] Add integration test: `test_docker_service_startup()` - verify all services start in order
- [ ] Add healthcheck test: `test_service_healthchecks()` - ensure endpoints respond

### 4. **Dense+BM25 Search Edge Cases**
- [ ] Test: empty results fallback
- [ ] Test: single result reranking
- [ ] Test: large result sets (100+ chunks)
- [ ] Test: BM25 scoring when no dense vector available

---

## Deployment Verification Checklist

Before merging to main, verify:

- [ ] **Docker Compose**: All services start with healthchecks passing
  ```bash
  docker compose up -d --build
  docker compose ps  # verify all services "healthy"
  ```

- [ ] **Harrier TEI Endpoint**: Responds to embedding requests
  ```bash
  curl http://localhost:8080/v1/embeddings -X POST \
    -H "Content-Type: application/json" \
    -d '{"input": ["hello world"], "model": "sentence-transformers/all-MiniLM-L6-v2"}'
  ```

- [ ] **Reranker TEI Endpoint**: Responds to reranking requests
  ```bash
  curl http://localhost:8081/rerank -X POST \
    -H "Content-Type: application/json" \
    -d '{"query": "test", "texts": ["hello"], "top_n": 1}'
  ```

- [ ] **Backend API**: Health check passes
  ```bash
  curl http://localhost:9090/health
  ```

- [ ] **Document Ingestion**: Upload a test document, verify status goes through Pending→Processing→Complete

- [ ] **RAG Search**: Query with test document, verify results are returned with reranking scores

---

## Tracking

- **PR**: #1 (Harrier embedding migration)
- **Commits Related**:
  - `3dbfb75`: Core migration
  - `22db7f7`: Documentation updates
  - `0234912`: Test cleanup + port fixes
  - `207dbcc`: .env.example PORT fix
- **Branch**: `claude/implement-swarm-8hDnh`
- **Review Status**: ✅ Approved for merge (all critical items resolved)

---

## Notes

- Reranking is **fully operational and critical to retrieval quality**. No action needed - system is working as designed.
- Sparse embedding column retention (Option A) is intentional for rollback safety.
- Embedding dimension migration is a known limitation that will be documented in upgrade guide.
- All test suite updates (stale test removal, new test additions) have been completed.

### PR DESCRIPTION
**Pull request by zaxbysauce**
## Summary

Closes #2, #12, #13, #14. Four issues representing significant tech debt and missing features in the KnowledgeVault RAG pipeline, shipped fully wired with no deferred work.

- **#14 — vault_id required on upload**: Both `POST /api/documents` and `POST /api/documents/upload` now use `vault_id: int = Query(...)` (Ellipsis, no default). Requests without `vault_id` return HTTP 422. Silent assignment to vault 1 is eliminated.
- **#12 — Parent-document retrieval (small-to-big)**: Chunks store a `parent_window_text` in metadata at ingest time (±3000 chars around the match, clamped to document bounds). When `PARENT_RETRIEVAL_ENABLED=true`, the RAG engine attaches the parent window to `RAGSource` objects and the prompt builder renders a `[[MATCH: ...]]` anchor within the window. Group-aware dedup (`PER_DOC_CHUNK_CAP=2`, `UNIQUE_DOCS_IN_TOP_K=5`) replaces the UID-strip dedup that collapsed strong chunks from the same document.
- **#13 — Ingestion integrity**: Three gaps closed: (1) atomic visibility — chunks from non-indexed files are filtered out of results via `indexed_file_ids` set; (2) safe re-upload — new generation chunks use hash-prefixed IDs and are inserted *before* old generation is deleted, eliminating the zero-chunk window; (3) ANN index lifecycle — `_maybe_rebuild_or_drop_vector_index` drops the IVF_PQ index below 256 rows and rebuilds when delete churn ≥ `INDEX_REBUILD_DELTA` (default 20%).
- **#2 — Embedding dimension migration tooling**: `scripts/migrate_embeddings.py` detects dimension mismatch, wipes LanceDB, resets file statuses to `pending`, with `--dry-run` and `--force` flags. `/api/health?deep=true` now returns `"stale_embeddings": true` when stored dimension ≠ `EMBEDDING_DIM`, with a WARNING log. README and `docs/release.md` include step-by-step upgrade instructions.

## What changed

| Area | Files |
|---|---|
| API — vault_id enforcement | `backend/app/api/routes/documents.py` |
| Health — stale embedding detection | `backend/app/api/routes/health.py` |
| Config — 7 new feature flags | `backend/app/config.py` |
| Chunking — parent window computation | `backend/app/services/chunking.py` |
| Document processor — parent window ingest, safe re-upload | `backend/app/services/document_processor.py` |
| Retrieval — group-aware dedup, visibility filter, parent window | `backend/app/services/document_retrieval.py` |
| Prompt builder — `[[MATCH:]]` rendering | `backend/app/services/prompt_builder.py` |
| RAG engine — indexed_file_ids lookup, parent window expansion | `backend/app/services/rag_engine.py` |
| Vector store — schema, ANN lifecycle, optimize, safe delete | `backend/app/services/vector_store.py` |
| Migration scripts | `backend/app/migrations/add_parent_window.py`, `backend/app/migrations/audit_vault_defaults.py` |
| Embedding migration CLI | `scripts/migrate_embeddings.py` |
| Docker compose — 7 new env vars wired | `docker-compose.yml` |
| Docs | `README.md`, `docs/release.md` |
| Tests (50 new) | `backend/tests/test_parent_retrieval.py` (23), `backend/tests/test_ingestion_integrity.py` (27) |

## Bugs fixed during review

A swarm-mode PR review identified and fixed three issues before merge:

1. **`compute_parent_windows` cursor bug** — `search_cursor` was advanced to `idx` (start of match) instead of `chunk_end` (end of match), so duplicate chunk text in a document would map all matching chunks to the first occurrence. Fixed: cursor now advances to `chunk_end`. A new test (`test_duplicate_chunk_text_advances_cursor`) covers this case and would have caught it.

2. **`delete_by_file` filter escaping** — Used manual double-quote escaping (`replace('"', '\\"')`) instead of `_lance_escape()` with single quotes, inconsistent with all other LanceDB filter expressions in the file. Fixed: now uses `_lance_escape()` and single-quote delimiters uniformly.

3. **docker-compose.yml missing 7 env vars** — The 7 new v0.2.0 config variables were not wired into the container, making them impossible to configure via `.env`. Fixed: all 7 added with `${VAR:-default}` pattern.

## Configuration

All new behaviour is feature-flagged with safe defaults:

| Variable | Default | Description |
|---|---|---|
| `PARENT_RETRIEVAL_ENABLED` | `false` | Enable small-to-big context expansion |
| `PARENT_WINDOW_CHARS` | `6000` | Total window size in characters |
| `NEW_DEDUP_POLICY` | `true` | Use group-aware dedup |
| `PER_DOC_CHUNK_CAP` | `2` | Max chunks per document in results |
| `UNIQUE_DOCS_IN_TOP_K` | `5` | Max distinct documents in result set |
| `INDEX_REBUILD_DELTA` | `0.2` | Delete churn fraction to trigger ANN rebuild |
| `REUPLOAD_SAFE_ORDER` | `true` | Insert new chunks before deleting old on re-upload |

## Breaking changes

**Embedding dimension change (Harrier migration):** If upgrading from BGE-M3 (768-dim) to Harrier (1024-dim), run the migration script before restarting:

```bash
# Dry run first
python scripts/migrate_embeddings.py --dry-run

# Apply
python scripts/migrate_embeddings.py
docker compose restart knowledgevault
```

See `README.md` → Upgrading and `docs/release.md` → v0.2.0 for full instructions.

## Known design trade-offs (not blocking)

- **ANN index churn check is per-call, not cumulative**: multiple small deletes (each below `INDEX_REBUILD_DELTA` individually) won't trigger a rebuild even if cumulative churn exceeds the threshold. The current design favours simplicity; operators can force a rebuild via `migrate_embeddings.py --force` if needed.
- **Parent windows not computed for schema/spreadsheet files**: `document_text` is empty for `.sql`/`.ddl` files and reconstructed text for spreadsheets, so `compute_parent_windows` is skipped. A debug log is emitted when `PARENT_RETRIEVAL_ENABLED=true` and the file type doesn't support windows.

## Test plan

- [x] `pytest tests/test_parent_retrieval.py tests/test_ingestion_integrity.py` — 50/50 passed
- [x] `POST /api/documents` without `vault_id` → HTTP 422 (verified via source: `Query(...)` no default)
- [x] `compute_parent_windows` — 9 test cases including duplicate chunk text cursor advancement
- [x] `_group_aware_dedup` — 6 test cases
- [x] Prompt builder `[[MATCH:]]` — 3 test cases
- [x] Migration dry-run idempotency — 3 test cases
- [x] ANN index lifecycle — 6 test cases
- [x] optimize() non-fatal — 2 test cases
- [x] Atomic visibility filter — 4 test cases
- [x] Safe re-upload — 4 test cases
- [x] vault_id required — 3 test cases
- [x] Audit migration — 8 test cases

https://claude.ai/code/session_011CsZuyJLVPhg4Yq1P75kPi

### CODE REVIEW

File: backend/app/services/document_processor.py [hunk 1]
```diff
@@ -892,13 +909,43 @@ async def process_file(
...
+                            and document_text
+                        ):
+                            chunk_metadata["parent_window_text"] = document_text[
+                                chunk.parent_window_start : chunk.parent_window_end
+                            ]
+
+                        # For safe re-upload, prefix chunk IDs with the new file hash
+                        # so old-generation chunks have distinguishable IDs (Issue #13)
+                        if settings.reupload_safe_order:
+                            record_id = f"{file_id}_{file_hash[:8]}_{chunk_scale}_{chunk.chunk_index}"
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply backend/app/services/document_processor.py:939]: Fixed in commit `3a1b4fa`.

Root cause confirmed: `expand_window()` was constructing legacy-format UIDs (`{file_id}_{scale}_{idx}`) while the vector store held new-format IDs (`{file_id}_{hash8}_{scale}_{idx}`). Since `get_chunks_by_uid` does exact `id IN (...)` matching, no adjacent chunks were found and window expansion silently returned only the original sources.

Fix:
- `filter_relevant()` now stores the actual stored chunk ID in `RAGSource.metadata["_chunk_id"]`
- Added `_extract_reupload_hash()` to detect the hash prefix from a source's stored ID
- `expand_window()` uses the hash prefix when building adjacent UIDs, falling back to legacy format for old-generation chunks
- Also fixed adjacent-chunk `RAGSource` construction to read `file_id` from the record rather than parsing it from the UID string (which was producing `"42_abc12345"` instead of `"42"` for new-format IDs)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: backend/app/services/vector_store.py [hunk 2]
```diff
@@ -273,6 +281,84 @@ async def _maybe_create_vector_index(self) -> None:
...
+                    "falling back to brute-force search.",
+                    current_rows,
+                    VECTOR_INDEX_MIN_ROWS,
+                )
+            except (OSError, RuntimeError, ValueError) as e:
+                logger.warning("Failed to drop IVF_PQ index after threshold crossed: %s", e)
+            return
+
+        # Case 2: Churn threshold exceeded — rebuild the index
+        if self._last_index_build_row_count > 0:
```
[the remark this answers is not in the document]
Comment (zaxbysauce) [reply backend/app/services/vector_store.py:350]: Fixed in commit `3a1b4fa`.

Root cause confirmed: `_last_index_build_row_count` is initialized to `0` in `__init__` and only set to a non-zero value when this process creates or rebuilds the index. On restart with an existing IVF_PQ index, the counter stays `0` and the `if _last_index_build_row_count > 0` guard in `_maybe_rebuild_or_drop_vector_index` permanently disables churn-based rebuilds.

Fix: in `init_table()`, after successfully opening an existing table, check `list_indices()` for an IVF_PQ index and seed `_last_index_build_row_count = await self.table.count_rows()`. Failure is non-fatal (debug log only) so it doesn't break startup on environments where `list_indices` is unavailable.

---
_Generated by [Claude Code](https://claude.ai/code)_
